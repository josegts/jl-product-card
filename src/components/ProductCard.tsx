import React from 'react';
import styles from '../styles/styles.module.css';
import { useProducts } from '../hooks/useProducts';
import { createContext, CSSProperties } from 'react';
import {
  InitialValues,
  OnChangeProps,
  ProductCardHandlers,
  ProductContextProps,
  ProductProps
} from '../interfaces/Interfaces';

export const ProductContext = createContext({} as ProductContextProps);
const { Provider } = ProductContext;

export interface Props {
  //children?: ReactElement | ReactElement[];
  children: (args: ProductCardHandlers) => JSX.Element;
  product: ProductProps;
  className?: string;
  style?: CSSProperties;
  onChange?: (args: OnChangeProps) => void;
  value?: number;
  initialValues?: InitialValues;
}

export const ProductCard = ({
  product,
  children,
  className,
  style,
  onChange,
  value,
  initialValues
}: Props) => {
  const { counter, increaseBy, isMaxCountReached, reset } = useProducts({
    product,
    onChange,
    value,
    initialValues
  });
  return (
    <Provider
      value={{
        counter,
        increaseBy,
        product,
        maxCount: initialValues?.maxCount
      }}
    >
      <div className={`${styles.productCard} ${className}`} style={style}>
        {children({
          count: counter,
          maxCount: initialValues?.maxCount,
          isMaxCountReached,
          increaseBy,
          reset,
          product
        })}
        {/** <ProductImage img={product.img} />
        <ProductTitle title={product.title} />
        <ProductButtons increaseBy={increaseBy} counter={counter} /> */}
      </div>
    </Provider>
  );
};
