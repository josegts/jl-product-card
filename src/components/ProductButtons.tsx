import React, { useCallback, useContext } from 'react';
import { ProductContext } from './ProductCard';
import styles from './../styles/styles.module.css';

export const ProductButtons = ({ className }: { className?: string }) => {
  // TODO: max count,
  const { increaseBy, counter, maxCount } = useContext(ProductContext);

  //TODO isMaxReached = useCallback, [ count, maxCount]
  const isMaxReached = useCallback(
    () => !!maxCount && counter === maxCount,
    [counter, maxCount]
  );

  console.log(isMaxReached());
  return (
    <div className={`${styles.buttonsContainer} ${className}`}>
      <button onClick={() => increaseBy(-1)} className={styles.buttonMinus}>
        -
      </button>
      <div className={`${styles.countLabel}`}>{counter}</div>
      <button
        onClick={() => increaseBy(+1)}
        className={`${styles.buttonAdd} ${isMaxReached() && styles.disabled}`}
      >
        +
      </button>
    </div>
  );
};
