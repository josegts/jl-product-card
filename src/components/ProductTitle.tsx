import React, { CSSProperties, useContext } from 'react';
import { ProductContext } from './ProductCard';
import styles from '../styles/styles.module.css';
export interface TitleProps {
  title?: string;
  className?: string;
  style?: CSSProperties;
}

export const ProductTitle = ({ title, style, className }: TitleProps) => {
  const { product } = useContext(ProductContext);
  return (
    <span style={style} className={`${styles.productDescription} ${className}`}>
      {title ? title : product.title}
    </span>
  );
};
