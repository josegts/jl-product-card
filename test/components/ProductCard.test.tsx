import React from 'react'
import renderer  from 'react-test-renderer'
import { ProductCard } from '../../src/components/ProductCard';
import { ProductTitle } from '../../src/components/ProductTitle';



describe('ProductCard', () => {
  const product1 = {
    id: '1',
    title: 'Coffee Mug with a nice picture',
    img: './coffee-mug.png'
  }

  test('should be show title with prodact data', () => {
    const wrapper = renderer.create(<ProductCard product={product1} >
      {() => (<ProductTitle/>)}
    </ProductCard>)

    expect(wrapper).toMatchSnapshot()
  })
})