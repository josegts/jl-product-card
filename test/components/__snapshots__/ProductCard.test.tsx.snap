// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ProductCard should be show title with prodact data 1`] = `
<div
  className="productCard undefined"
>
  <span
    className="productDescription undefined"
  >
    Coffee Mug with a nice picture
  </span>
</div>
`;
